### DISCUSSION TITLE
survey เกมเขียนโปรแกรม

### DISCUSSION THREAD

**Discussion by ppbasleng** [Logical-Learning-Game/Document#34] [Show and tell]
1.หา List ของชื่อเกมเขียนโปรแกรมที่น่าจะเอามาเป็นตัวอย่างได้
2.ไปลองเล่นเกมจากใน list ที่หามา ศึกษาดูแนวทางน่าจะปรับมาใช้กันเกมได้
3.สรุปผลการ survey (อธิบายประเภทของแต่ละเกมที่เจอ เสนอไอเดียที่น่าจะเอามาใช้กับเกมได้)

**Comment by Chuckboliver** [comment]
# SHENZHEN I/O
เกมสร้างระบบ Embedded System และแก้ปัญหาโดยการเขียนโปรแกรม (ใช้ภาษาคล้ายๆ Assembly)
https://www.youtube.com/watch?v=Av3ZZAiA0n4

## Example
โจทย์ให้เราสร้าง Pulse Generator โดยผลลัพธ์ที่ถูกต้องจะต้องเป็นไปตาม Timing Diagram ด้านล่าง

![Image](https://user-images.githubusercontent.com/43881371/184493771-eb1228a2-5464-40a3-9bf5-32ddfe0bd359.png)

เราสามารถผ่านโจทย์นี้ได้หลายวิธี ซึ่งหลังจากผ่านด่านเกมก็จะคำนวณว่าเราผ่านโดยใช้เงินในการสร้างไปเท่าไร, ใช้ Power ไปเท่าไร, เขียนโค้ดรวมแล้วกี่บรรทัด
![image](https://user-images.githubusercontent.com/43881371/184494051-c95f9078-4672-4c5e-8f56-d32b50eb8d12.png)


## Idea ที่น่าจะเอามาปรับใช้กับ Project เราได้:
* สามารถ Debug โดยการ Step code ไปทีละบรรทัดหรือ Step ไปทีละ 1 รอบการทำงานได้
* ระหว่าง Simulate code อยู่สามารถ Run/Pause ได้
* แต่ละชิ้นส่วนที่เอามาใช้ในการแก้ปัญหามี Cost
* สามารถเขียนโค้ดในหลายๆส่วนที่สามารถทำงานพร้อมกันได้

**Comment by ppbasleng** [comment]
# Code Combat
เกมเขียนโปรแกรมสไตล์ rpg โดยการเขียน command ในการสั่งให้ตัวละครทำ action ตาม
## Example


![Image](https://user-images.githubusercontent.com/54877995/184494991-6b34bfa5-44ec-474a-8fc1-13a86be8f82e.png)

ด้านซ้ายหน้าจอจะเป็นส่วนแสดงผลการทำงาน ซึ่งก็คือด่าน
ด้านขวาหน้าจอจะเป็น console สำหรับเขียนโปรแกรม
## Link วีดิโอ walkthrough จาก youtube
https://www.youtube.com/playlist?list=PLbsvRhEyGkKdXNs-z1SYaWmMLMgVENRgI
(เนื่องจากในด่านหลัง ๆ ของเกม ผู้ใช้จะต้องทำการ subscription จึงไม่สามารถเห็นระบบเกมโดยรวมได้ )
## Idea ที่น่าจะเอามาปรับใช้ได้
- ระบบเกมพื้นฐานโดยรวมที่ใช้ command ในการสั่งให้ตัวละครทำ action
- การเคลื่อนที่ในทิศทางเดียวกันสามารถใส่ parameter ในการควบคุมจำนวนครั้งได้
- มีการใช้ loop ในเกมนี้จะเป็นการใช้ while true loop เพื่อไม่ให้เกมมีความซับซ้อนมาเกินไป
- มีการเล่นเรื่องของ variable ภายในเกม (ค้นหาศัตรูใกล้ๆเป็น enemy1 แล้วโจมตี enemy1)

**Comment by Chuckboliver** [comment]
# EXAPUNKS
เกมที่เราจะรับบทเป็น Hacker ที่จะต้องเขียนโค้ดควบคุมไวรัสคอมพิวเตอร์เพื่อเข้าไปแก้ไขเปลี่ยนแปลงระบบต่างๆในเครื่องเป้าหมาย
https://www.blognone.com/node/106197

## Example
![image](https://user-images.githubusercontent.com/43881371/184496048-d6727f25-b5ae-4693-8148-4e387bfee296.png)

## Idea ที่น่าจะเอามาปรับใช้ได้
* มีมินิเกมให้เล่นฆ่าเวลาเมื่อคิดปริศนาหลักไม่ออก
![image](https://user-images.githubusercontent.com/43881371/184496306-585e0ab6-97d1-45e8-a253-c0d52e94bb63.png)
* สามารถเขียนโค้ดได้ยาวเท่าใดก็ได้เพื่อให้โปรแกรมทำงานถูกต้องและดำเนินเนื้อเรื่องต่อ เพียงแค่แลกมากับการไม่สามารถอัพโหลดคะแนนขึ้นตารางเท่านั้น
* ผู้เล่นสามารถสร้างด่านขึ้นมาเป็นของตัวเองโดยใช้ภาษา Javascript
